game/world/triggers/cscamera.h

```diff
@@ -9,8 +9,40 @@ class CsCamera : public AbstractTrigger {
   public:
     CsCamera(Vob* parent, World& world, const phoenix::vobs::cs_camera& data, Flags flags);
 
+  private:
+    struct KeyFrame {
+      float         time  = 0;
+      Tempest::Vec3 c[4]  = {};
+      float arcLength(float t0 = 0, float t1 = 1);
+      };
+
+    struct CamSpline {
+      float                 t;
+      float                 dist;
+      float                 c[3]     = {};
+      std::vector<KeyFrame> keyframe = {};
+      uint32_t size() { return uint32_t(keyframe.size()); }
+      float    nextDist(float t);
+      KeyFrame& operator [](const size_t n) { return keyframe[n]; }
+      };
+
     void onTrigger(const TriggerEvent& evt) override;
+    void onUntrigger(const TriggerEvent& evt) override;
+    void tick(uint64_t dt) override;
 
-  private:
+    void init(const phoenix::vobs::cs_camera& cam);
+    void clear();
+
+    auto position(CamSpline& cs) -> Tempest::Vec3;
+    auto spin(const Tempest::Vec3& d) -> Tempest::PointF;
+
+    static inline uint8_t   activeEvents;
```

Comment:
> Also does it make sense to allow more than one CsCamera at a time?

There is a bug that makes two occur at the same time

```
Trigger: EVT_LEFT_UP_CAMERA_01
Trigger: EVT_LEFT_DOWN_CAMERA_01
Untrigger: EVT_LEFT_UP_CAMERA_01
Untrigger: EVT_LEFT_DOWN_CAMERA_01
```

In vanilla the up event is triggered first, the down event if the levers are used a second time at Irdorath switch riddle location. The variable is only used to guard against it.

Reply:
If this suppose to be like "fist-in" scheme, than having a pointer to active CsCamera can be more convinient:
```
CsCamera* WorldObjects::currentCs = nullptr;
```

Reply:
> Also does it make sense to allow more than one CsCamera at a time? Any idea on how vanilla react to it?

I tested a bit and during timedemo portal opening sequence can be triggered. Timedemo is not continued after finish. This would suggest a new cscamera disables the previous.

> If this suppose to be like "fist-in" scheme,

`Up` event is always first and `zSpy` shows only one `OnTrigger()` event in vanilla. There must be something else before that decides which event to take.

>  a pointer to active CsCamera can be more convinient:

Makes sense with the above observation. Should this be accessed by getter and setter methods through `World` ?

Reply:
> Should this be accessed by getter and setter methods through World ?

Yep, that is a good option; also in general - no need to ask such minor details ;)